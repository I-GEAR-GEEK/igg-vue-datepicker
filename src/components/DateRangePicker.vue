<template>
  <DatePicker
    ref="datepicker"
    v-model="dateRange"
    range
    format="DD/MM/YYYY"
    :shortcuts="shortcuts"
    :placeholder="placeholder"
    range-separator=" - "
    :lang="lang"
    :disabled-date="disabledDate"
    style="width: 100%"
  />
</template>

<script>
import DatePicker from 'vue2-datepicker'
import 'vue2-datepicker/index.css'
import moment from 'moment'
import { LANG } from '@/assets/constant/datepicker'

export default {
  components: {
    DatePicker
  },
  props: {
    value: {
      type: Array,
      default: () => ([])
    },
    placeholder: {
      type: String,
      default: ''
    },
    disabledDate: {
      type: Function,
      default: () => {}
    }
  },
  data () {
    return {
      lang: LANG,
      dateRange: this.value,
      shortcuts: [
        {
          text: 'เมื่อวานนี้',
          onClick: () => this.setDateAgo(1)
        },
        {
          text: '7 วันที่ผ่านมา',
          onClick: () => this.setDateAgo(7)
        },
        {
          text: '14 วันที่ผ่านมา',
          onClick: () => this.setDateAgo(14)
        },
        {
          text: '30 วันที่ผ่านมา',
          onClick: () => this.setDateAgo(30)
        },
        {
          text: 'สัปดาห์นี้',
          onClick: () => this.setDateWeek()
        },
        {
          text: 'สัปดาห์ที่แล้ว',
          onClick: () => this.setDateLastWeek()
        },
        {
          text: 'เดือนนี้',
          onClick: () => this.setDateMonth()
        },
        {
          text: 'เดือนที่แล้ว',
          onClick: () => this.setDateLastMonth()
        }
      ]
    }
  },
  watch: {
    value (val) {
      this.dateRange = val
    },
    dateRange (val) {
      this.$refs.datepicker.closePopup()
      this.$emit('input', val)
    }
  },
  methods: {
    setDateAgo (day) {
      this.dateRange = [new Date(moment().subtract(day, 'days').format('YYYY-MM-DD')), new Date(moment().format('YYYY-MM-DD'))]
    },
    setDateWeek () {
      this.dateRange = [new Date(moment().startOf('week').format('YYYY-MM-DD')), new Date(moment().endOf('week').format('YYYY-MM-DD'))]
    },
    setDateLastWeek () {
      this.dateRange = [new Date(moment().startOf('week').subtract(7, 'days').format('YYYY-MM-DD')), new Date(moment().endOf('week').subtract(7, 'days').format('YYYY-MM-DD'))]
    },
    setDateMonth () {
      this.dateRange = [new Date(moment().startOf('month').format('YYYY-MM-DD')), new Date(moment().endOf('month').format('YYYY-MM-DD'))]
    },
    setDateLastMonth () {
      this.dateRange = [new Date(moment().subtract(1, 'month').startOf('month').format('YYYY-MM-DD')), new Date(moment().subtract(1, 'month').endOf('month').format('YYYY-MM-DD'))]
    }
  }
}
</script>

<style>
.mx-datepicker-sidebar {
  float: right;
  width: 120px;
}
.mx-datepicker-sidebar+.mx-datepicker-content {
  margin-left: 0;
  border-left: 0px solid #e8e8e8;
  margin-right: 120px;
  border-right: 1px solid #e8e8e8;
}
</style>
