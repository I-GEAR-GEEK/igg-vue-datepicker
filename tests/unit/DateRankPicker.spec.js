
import { shallowMount } from '@vue/test-utils'
import DateRangePicker from '@/components/DateRangePicker.vue'

global.Date.now = jest.fn(() => new Date('2021-07-02T00:00:00'))

describe('DateRangePicker.vue', () => {
  it('should match with snapshot', () => {
    const wrapper = shallowMount(DateRangePicker, {
      placeholder: 'placeholder'
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should match with value', async () => {
    const wrapper = await shallowMount(DateRangePicker, {
      propsData: {
        value: [new Date('2021-07-01T00:00:00.000Z'), new Date('2021-07-02T00:00:00.000Z')]
      }
    })

    wrapper.vm.$refs.datepicker.closePopup = jest.fn()
    const expectResult = [new Date('2021-07-01T00:00:00.000Z'), new Date('2021-07-02T00:00:00.000Z')]
    expect(wrapper.vm.$data.dateRange).toEqual(expectResult)
  })
  it('should match with yesterday', async () => {
    const wrapper = await shallowMount(DateRangePicker)
    wrapper.vm.$refs.datepicker.closePopup = jest.fn()
    await wrapper.vm.setDateAgo(1)

    const expectResult = [new Date('2021-07-01T00:00:00.000Z'), new Date('2021-07-02T00:00:00.000Z')]
    expect(wrapper.vm.$data.dateRange).toEqual(expectResult)
  })
  it('should match with current week', async () => {
    const wrapper = await shallowMount(DateRangePicker)
    wrapper.vm.$refs.datepicker.closePopup = jest.fn()
    await wrapper.vm.setDateWeek()

    const expectResult = [new Date('2021-06-27T00:00:00.000Z'), new Date('2021-07-03T00:00:00.000Z')]
    expect(wrapper.vm.$data.dateRange).toEqual(expectResult)
  })
  it('should match with last week', async () => {
    const wrapper = await shallowMount(DateRangePicker)
    wrapper.vm.$refs.datepicker.closePopup = jest.fn()
    await wrapper.vm.setDateLastWeek()

    const expectResult = [new Date('2021-06-20T00:00:00.000Z'), new Date('2021-06-26T00:00:00.000Z')]
    expect(wrapper.vm.$data.dateRange).toEqual(expectResult)
  })
  it('should match with current month', async () => {
    const wrapper = await shallowMount(DateRangePicker)
    wrapper.vm.$refs.datepicker.closePopup = jest.fn()
    await wrapper.vm.setDateMonth()

    const expectResult = [new Date('2021-07-01T00:00:00.000Z'), new Date('2021-07-31T00:00:00.000Z')]
    expect(wrapper.vm.$data.dateRange).toEqual(expectResult)
  })
  it('should match with last month', async () => {
    const wrapper = await shallowMount(DateRangePicker)
    wrapper.vm.$refs.datepicker.closePopup = jest.fn()
    await wrapper.vm.setDateLastMonth()

    const expectResult = [new Date('2021-06-01T00:00:00.000Z'), new Date('2021-06-30T00:00:00.000Z')]
    expect(wrapper.vm.$data.dateRange).toEqual(expectResult)
  })
})
