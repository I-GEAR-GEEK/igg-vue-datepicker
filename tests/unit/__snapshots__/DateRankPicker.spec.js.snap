// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePicker.vue should match with snapshot 1`] = `
<datepicker-stub value="" defaultvalue="Mon Jul 12 2021 00:00:00 GMT+0700 (Indochina Time)" disableddate="() => {}" type="date" getclasses="function _default() {
          return [];
        }" titleformat="YYYY-MM-DD" format="DD/MM/YYYY" timetitleformat="YYYY-MM-DD" disabledtime="function _default() {
          return false;
        }" hourstep="1" minutestep="1" secondstep="1" scrollduration="100" valuetype="date" range="true" rangeseparator=" - " lang="[object Object]" placeholder="" editable="true" clearable="true" prefixclass="mx" inputattr="[object Object]" appendtobody="true" popupstyle="[object Object]" confirmtext="OK" shortcuts="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" style="width: 100%;"></datepicker-stub>
`;
